clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -709,10 +709,10 @@ public void onLeavePrepare() {
 
         if (subscriptions.hasAutoAssignedPartitions() && !droppedPartitions.isEmpty()) {
             final Exception e;
-            if (generation() != Generation.NO_GENERATION) {
-                e = invokePartitionsRevoked(droppedPartitions);
-            } else {
+            if (generation() == Generation.NO_GENERATION || rebalanceInProgress()) {
```

Comment:
@ableegoldman @mjsax This is another bug I found while trouble-shooting the system test failures (dates before the cleanup): when we got a task-migrated exception, and then enforce a rebalance, we call `unsubscribe` which would trigger `onLeavePrepare`, here if it was from task-migrated then it is likely that we are already undergoing a rebalance and in that case we should lose the tasks instead of revoke them since otherwise, we would still try to `commit` which would fail with a RebalanceInProgress exception.

Reply:
Good catch. Also it's pretty unfortunate that we can only trigger a rebalance from outside the client by unsubscribing and closing/suspending the entire assignment...this limits the usefulness of KIP-429 during version probing upgrades.

It also has some implications for the "rebalances are cheap" assumption of KIP-441. Would be better phrased as "rebalances are cheap, except for the member who triggers them".

Reply:
Yeah ... maybe we should consider adding a new API to consumer to `rejoin` the group, in a cheaper way.

Reply:
I agree :) -- would be happy to write up a small KIP for it and kick off discussion